interfaces/IPower.h

```diff
@@ -35,10 +35,21 @@ namespace Exchange {
             PowerOff = 6, // S5.
         };
 
+        enum PCPhase : uint8_t {
+            Before = 1,
+            After  = 2
+        };
+
+        struct PowerState {
+            PCState origin;
+            PCState destination;
+            PCPhase phase;
+        };
+
         struct EXTERNAL INotification : virtual public Core::IUnknown {
             enum { ID = ID_POWER_NOTIFICATION };
 
-            virtual void StateChange(const PCState) = 0;
+            virtual void StateChange(const PowerState &state) = 0;
```

Comment:
As this inherits from IUnknown it is allowed to pass process boundaries and even, in R3) host boundaries, hence why we are currently checking for enums for size.
Because it can pass also host boundaries, it could potentially pass from a 32 bit host to a 64 bit host. This, in structs might lead to incompatiblitlies (packing/alignment of members in the struct).
To avoid these complications, Suggest to send out the values as seperate elements, so: 
virtual void StateChange(const PCState origin, const PCState destination, const PCPhase phase) = 0;

The issue above is a real issue and should be addressed. As some food for thought, not mandatory but just a suggestion, if you need this interface to be more descriptive (based on what I understand of the PCPhase member) is to change the interface to explicit methods like:

virtual uint32_t RequestChange(const PCState from, const PCState to) = 0;
virtual void StateChange(const PCState from, const PCState to) = 0;

Where the first method allows you to "block" a state change if needed, and the second shows that it is completed. If this makes sense....

Reply:
Agree with your suggestion to send out the values as separate elements. 

But the approach using `RequestChange()` and `StateChange()`  would probably require even more changes for implementing `RequestChange()`, as the callee could have the ability to block the transition. (In case a callee wants to block the transition, the transition should be blocked and the observers should be notified that the transition is aborted.) So would prefer to use `virtual void StateChange(const PCState origin, const PCState destination, const PCPhase phase) = 0;`.